components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +72,154 @@ impl MutationObserver {
+                    // Step 3.1
+                    if !interestedObservers.contains(&registered_observer) {
+                        interestedObservers.push(JS::from_ref(registered_observer));
+                    }
+                    // Step 3.2
+                    let condition: bool = (given_name == "attribute" &&
+                        registered_observer.attribute_old_value.get() == true) ||
+                            (given_name == "characterData" &&
+                                registered_observer.character_data_old_value.get() == true);
+                    if condition {
+                        pairedStrings.push(DOMString::from(old_value));
+                    }
+                }
+            }
+        }
+        // Step 4
+        let mut i = 0;
+        for observer in &interestedObservers {
+            //Step 4.1
+            let mut record = &MutationRecord::new(DOMString::from(&*given_name), target);
+            //Step 4.2
+            if given_name != "" && given_namespace != "" {
+                record.SetAttributeName(DOMString::from(&*given_name));
+                record.SetAttributeNamespace(DOMString::from(&*given_namespace));
+            }
+            //Step 4.3-4.6- TODO currently not relevant to mutation records for attribute mutations
+            //Step 4.7
+            if pairedStrings[i] != "" {
+                record.SetoldValue(pairedStrings[i].clone());
+            }
+            i = i + 1;
+        }
+        // Step 5
+        let enqueuedMutationCallback = EnqueuedMutationCallback { callback: self.callback,
+            pipeline: target.global().pipeline_id() };
+        let compoundMicrotask = Microtask::Mutation(enqueuedMutationCallback);
+        MutationObserver::queueMutationObserverCompoundMicrotask(compoundMicrotask);
+    }
+
+}
+
+impl MutationObserverMethods for MutationObserver {
+    /// https://dom.spec.whatwg.org/#dom-mutationobserver-observe
+    /// MutationObserver.observe method
+    fn Observe(&self, target: &Node, options: &MutationObserverInit) -> Fallible<()> {
+        // Step 1: If either options’ attributeOldValue or attributeFilter is present and
+        // options’ attributes is omitted, set options’ attributes to true.
+        if (options.attributeOldValue.is_some() || options.attributeFilter.is_some()) && options.attributes.is_none() {
+//            options.attributes = Some(true);
+        }
+        // Step2: If options’ characterDataOldValue is present and options’ characterData is omitted,
+        // set options’ characterData to true.
+        if options.characterDataOldValue.is_some() && options.characterData.is_some() {
+//            options.characterData = Some(true);
+        }
+        // Step3: If none of options’ childList, attributes, and characterData is true, throw a TypeError.
+        if !options.childList && !options.attributes.unwrap() && !options.characterData.unwrap() {
```

Comment:
Prefer `unwrap_or(false)` for all of these checks, which will provide a default value if the dictionary member is not present rather than making Servo crash.